recipe-client-addon/lib/ShieldPreferences.jsm

```diff
@@ -0,0 +1,133 @@
+      case NS_PREFBRANCH_PREFCHANGE_TOPIC_ID:
+        if (data === FHR_UPLOAD_ENABLED_PREF) {
+          const fhrUploadEnabled = Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF);
+          Services.prefs.setBoolPref(OPT_OUT_STUDIES_ENABLED_PREF, fhrUploadEnabled);
+        }
+        break;
+    }
+  },
+
+  /**
+   * Injects the opt-out-study preference checkbox into about:preferences and
+   * handles events coming from the UI for it.
+   */
+  OptOutStudyCheckbox: class {
+    constructor(document) {
+      this.document = document;
+      this.refs = {};
+
+      const container = this.refs.container = document.createElementNS(XUL_NS, "vbox");
+      container.classList.add("indent");
+
+      const hContainer = this.refs.hContainer = document.createElementNS(XUL_NS, "hbox");
+      hContainer.setAttribute("align", "center");
+      container.appendChild(hContainer);
+
+      const checkbox = this.refs.checkbox = document.createElementNS(XUL_NS, "checkbox");
+      checkbox.setAttribute("id", "optOutStudiesEnabled");
+      checkbox.setAttribute("label", "Allow Firefox to install and run studies");
+      checkbox.setAttribute("preference", OPT_OUT_STUDIES_ENABLED_PREF);
+      checkbox.setAttribute("checked", Services.prefs.getBoolPref(OPT_OUT_STUDIES_ENABLED_PREF));
+      checkbox.setAttribute("disabled", !Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF));
+      hContainer.appendChild(checkbox);
+
+      const viewStudies = this.refs.viewStudies = document.createElementNS(XUL_NS, "label");
+      viewStudies.setAttribute("id", "viewShieldStudies");
+      viewStudies.textContent = "View Firefox Studies";
+      viewStudies.classList.add("learnMore", "text-link");
+      viewStudies.addEventListener("click", this);
+      hContainer.appendChild(viewStudies);
+
+      // <prefrence> elements for prefs that we need to monitor while the page is open.
+      const optOutPref = this.refs.optOutPref = document.createElementNS(XUL_NS, "preference");
+      optOutPref.setAttribute("id", OPT_OUT_STUDIES_ENABLED_PREF);
+      optOutPref.setAttribute("name", OPT_OUT_STUDIES_ENABLED_PREF);
+      optOutPref.setAttribute("type", "bool");
+
+      // Weirdly, FHR doesn't have a <preference> element on the page, so we create it.
+      const fhrPref = this.refs.fhrPref = document.createElementNS(XUL_NS, "preference");
+      fhrPref.setAttribute("id", FHR_UPLOAD_ENABLED_PREF);
+      fhrPref.setAttribute("name", FHR_UPLOAD_ENABLED_PREF);
+      fhrPref.setAttribute("type", "bool");
+      fhrPref.addEventListener("change", this);
+    }
+
+    inject() {
+      const parent = this.document.getElementById("submitHealthReportBox").closest("vbox");
+      parent.appendChild(this.refs.container);
+
+      const preferences = this.document.getElementById("privacyPreferences");
+      preferences.appendChild(this.refs.optOutPref);
+      preferences.appendChild(this.refs.fhrPref);
+    }
+
+    remove() {
+      this.refs.container.remove();
```

Comment:
I talked to @Mossop briefly about this and he agrees that the cycle collector handles this, and debug build tests should catch this, so we're probably fine. I'd prefer to not add extra cleanup handling if possible, shall we leave this as-is (besides removing the `remove` method, of course)?